presto-orc/src/main/java/com/facebook/presto/orc/reader/SliceDictionarySelectiveReader.java

```diff
@@ -0,0 +1,610 @@
+            int dataToSkip = presentStream.countBitsSet(items);
+            if (inDictionaryStream != null) {
+                inDictionaryStream.skip(dataToSkip);
+            }
+            dataStream.skip(dataToSkip);
+        }
+        else {
+            if (inDictionaryStream != null) {
+                inDictionaryStream.skip(items);
+            }
+            dataStream.skip(items);
+        }
+    }
+
+    @Override
+    public int[] getReadPositions()
+    {
+        return outputPositions;
+    }
+
+    @Override
+    public Block getBlock(int[] positions, int positionCount)
+    {
+        checkArgument(outputPositionCount > 0, "outputPositionCount must be greater than zero");
+        checkState(outputRequired, "This stream reader doesn't produce output");
+        checkState(positionCount <= outputPositionCount, "Not enough values");
+        checkState(!valuesInUse, "BlockLease hasn't been closed yet");
+
+        if (allNulls) {
+            return new RunLengthEncodedBlock(outputType.createBlockBuilder(null, 1).appendNull().build(), outputPositionCount);
+        }
+
+        if (positionCount == outputPositionCount) {
+            return new DictionaryBlock(positionCount, dictionaryBlock, values);
+        }
+
+        int[] valuesCopy = new int[positionCount];
+
+        int positionIndex = 0;
+        int nextPosition = positions[positionIndex];
+        for (int i = 0; i < outputPositionCount; i++) {
+            if (outputPositions[i] < nextPosition) {
+                continue;
+            }
+            assert outputPositions[i] == nextPosition;
+            valuesCopy[positionIndex] = this.values[i];
+            positionIndex++;
+            if (positionIndex >= positionCount) {
+                break;
+            }
+            nextPosition = positions[positionIndex];
+        }
+
+        return new DictionaryBlock(positionCount, dictionaryBlock, valuesCopy);
+    }
+
+    @Override
+    public BlockLease getBlockView(int[] positions, int positionCount)
+    {
+        checkArgument(outputPositionCount > 0, "outputPositionCount must be greater than zero");
+        checkState(outputRequired, "This stream reader doesn't produce output");
+        checkState(positionCount <= outputPositionCount, "Not enough values");
+        checkState(!valuesInUse, "BlockLease hasn't been closed yet");
+
+        if (positionCount < outputPositionCount) {
+            compactValues(positions, positionCount);
+        }
+        return newLease(new DictionaryBlock(positionCount, dictionaryBlock, values));
+    }
+
+    private void compactValues(int[] positions, int positionCount)
+    {
+        if (outputPositionsReadOnly) {
+            outputPositions = Arrays.copyOf(outputPositions, outputPositionCount);
+            outputPositionsReadOnly = false;
+        }
+        int positionIndex = 0;
+        int nextPosition = positions[positionIndex];
+        for (int i = 0; i < outputPositionCount; i++) {
```

Comment:
in direct reader you wrote this loop differently; any particular reason for not having both be the same?

Reply:
yes there I m copying the data array using, here I just used it like the other readers which deal with only values and nulls